File: Array/MinimumPathSum.java
import java.util.Queue;
import java.util.LinkedList;

// Source: https://leetcode.com/problems/minimum-path-sum/#/description
// Approach: Implement Dijkstra with a queue to find a path from top to bottom. Treat the grid as an actual graph
// This is just a pseudo solution. Complele solution is in the python file

public class Solution {
    
    public class Input {
        public int[][] grid;
        public int length;
        public int width;
        public void Input(int[][] grid){
            this.grid = grid;
            this.length = grid.length;
            this.width = grid[0].length;
        }
        public LinkedList<LinkedList<Integer>> neighbors(LinkedList<Integer> coord){
            LinkedList<LinkedList<Integer>> neighbors = new LinkedList<LinkedList<Integer>>();
            if (coord.get(0) + 1 < this.length){
                neighbors.add(new LinkedList<Integer>(Arrays.asList(coord.get(0) + 1, coord.get(1))));
            }
            if(coord.get(1) + 1 < this.width){
                neighbors.add(new LinkedList<Integer>(Arrays.asList(coord.get(0), coord.get(1) + 1)));
            }
            return neighbors;
        }
    }
    public int minPathSum(int[][] origin) {
        if(origin.length == 1 && origin[0].length == 1){
            return origin[0][0];
        }
        
        Queue<Integer> queue = new LinkedList<>();
        Input input = new Input(origin);
        queue.add(origin[0][0]);
        int[][] best = new int[input.length][input.width];
        for(int i=0; i<input.length; i++) {
            for(int j=0; j<input.width; j++) {
                best[i][j] = Integer.MAX_VALUE;
            }
        }
        
        while (queue.size() != 0){
            LinkedList<Integer> node = queue.remove();
            LinkedList<LinkedList<Integer>> neighbors = input.neighbors(node);
            for(LinkedList<Integer> neighbor : neighbors){
                int current = best[neighbor.get(0)][neighbor.get(1)];
                int parent = best[node.get(0)][node.get(1)];
                int nodeVal = origin[neighbor.get(0)][neighbor.get(1)];
                int newVal = parent + nodeVal;
                if (parent + nodeVal > current){ //reset
                    newVal = current;
                }
                best[neighbor.get(0)][neighbor.get(1)] = newVal;
                queue.add(neighbor);
            }
            return best[input.length - 1][input.wdith -1];
        }
    }
}

File: Array/MinimumPathSum.py
from collections import deque

class Input(object):
    """docstring for Input"""
    def __init__(self, matrix):
        self.matrix = matrix
        self.length = len(matrix)
        self.width = len(matrix[0])

    #coord: [y,x]
    def neighbors(self, coord):
        neighbors = []
        if coord[0] + 1 < self.length:
            neighbors.append([coord[0] + 1, coord[1]])
        if coord[1] + 1 < self.width:
            neighbors.append([coord[0], coord[1] + 1])
        return neighbors

class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        origin = grid
        if len(origin) == 1 and len(origin[0]) == 1:
            # print( origin[0][0])
            return origin[0][0]
        
        queue = deque()
        
        input = Input(origin)
        queue.append([0,0])
        
        best = [[99999 for y in range(0,input.width)] for i in range(0,input.length)]
        best[0][0] = origin[0][0]
        while len(queue) != 0:
            node = queue.popleft()
            neighbors = input.neighbors(node) # get neighbors of node
            for neighbor in neighbors:
                # if best[neighbor[0]][neighbor[1]] != -1:
                current = best[neighbor[0]][neighbor[1]]
                parent = best[node[0]][node[1]]
                nodeVal = origin[neighbor[0]][neighbor[1]]
                newVal = current if parent + nodeVal > current else  parent + nodeVal
                # print("Index", neighbor[0],neighbor[1], "current", current,"parent", parent)
                best[neighbor[0]][neighbor[1]] = newVal
                queue.append(neighbor)
        # print(best[input.length - 1][input.width - 1])
        return best[input.length - 1][input.width - 1]
        




